pyquil/quil.py

```diff
@@ -213,17 +213,20 @@ def inst(self, *instructions: InstructionDesignator) -> "Program":
                 else:
                     self.inst(" ".join(map(str, instruction)))
             elif isinstance(instruction, str):
-                self.inst(RSProgram(instruction.strip()))
+                self.inst(RSProgram.from_string(instruction.strip()))
             elif isinstance(instruction, AbstractInstruction):
-                self.inst(RSProgram(instruction.out()))
+                self.inst(RSProgram.from_string(instruction.out()))
             elif isinstance(instruction, Program):
                 self.inst(instruction._program)
             elif isinstance(instruction, RSInstruction):
                 self._program.add_instruction(instruction)
             elif isinstance(instruction, RSProgram):
                 self._program += instruction
             else:
-                self.inst(str(instruction))
+                try:
+                    self.inst(str(instruction))
+                except Exception:
+                    raise TypeError("Invalid instruction: {}".format(instruction))
```

Comment:
Is your intent to keep this after the baseclass work is done, or is this temporary?

The exception rewriting is a good idea.

Reply:
I was actively working on that today, and it turned out to be bit more thorny than it seemed.

I wrote up a [more detailed issue in quil-rs](https://github.com/rigetti/quail-rs/issues/148) describing why that is and some potential options.

For now, I've introduced a workaround, see my new comment [here](https://github.com/rigetti/pyquil/pull/1518/files#r1117803869)